crates/subspace-farmer/src/dsn.rs

```diff
@@ -1,15 +1,99 @@
-use std::ops::Range;
+use std::ops::{ControlFlow, Range};
+use std::sync::{Arc, Mutex};
 
-use futures::Stream;
-use subspace_core_primitives::PieceIndexHash;
+use futures::{Stream, StreamExt};
+
+use subspace_core_primitives::{FlatPieces, PieceIndex, PieceIndexHash, PublicKey, Sha256Hash};
+use subspace_solving::PieceDistance;
 
 use crate::PiecesToPlot;
 
+// TODO: Should it be a distinct type which implements arithmetics operations?
+pub type PieceIndexHashNumber = PieceDistance;
+
+/// Options for syncing
+pub struct SyncOptions {
+    /// The size of request in pieces which we should request on average
+    pub pieces_per_request: u64,
+    /// The size of range which we request initially
+    pub initial_range_size: PieceIndexHashNumber,
+    /// The size of range at which we should stop syncing
+    pub max_range_size: PieceIndexHashNumber,
+    /// Address of our plot
+    pub address: PublicKey,
+}
+
 #[async_trait::async_trait]
 pub trait DSNSync {
     type Stream: Stream<Item = PiecesToPlot>;
 
+    /// Get pieces from the network which fall within the range
     async fn get_pieces(&mut self, range: Range<PieceIndexHash>) -> Self::Stream;
+
+    async fn sync<OP>(&mut self, options: SyncOptions, on_pieces: OP) -> anyhow::Result<()>
```

Comment:
Not only algorithm here isn't very clear to me, I'm not sure why this needs to be a part of DSN sync, it seems like you're mixing APIs provided by networking (DSN) and farmer's implementation detail of the sync process.

Reply:
Yeah, makes sense. I just thought at first that it will be owned solely by the farmer. Will move it to some function in this module